httpserver/handlers.go

```diff
@@ -83,19 +87,101 @@ func (server *Server) verify(ctx context.Context, w http.ResponseWriter, r *http
 				Key:   subject,
 				Value: fromVerifyResult(result),
 			})
+			logrus.Debugf("verification: execution time for image %s: %dms", subject, time.Since(routineStartTime).Milliseconds())
 		}(utils.SanitizeString(subject))
 	}
 	wg.Wait()
+	logrus.Debugf("verification: execution time for request: %dms", time.Since(startTime).Milliseconds())
+	return sendResponse(&results, "", w, http.StatusOK, false)
+}
+
+func (server *Server) mutate(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	startTime := time.Now()
+	sanitizedMethod := utils.SanitizeString(r.Method)
+	sanitizedURL := utils.SanitizeURL(*r.URL)
+	logrus.Infof("start request %s %s", sanitizedMethod, sanitizedURL)
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return fmt.Errorf("unable to read request body: %v", err)
+	}
+
+	// parse request body
+	var providerRequest externaldata.ProviderRequest
+	err = json.Unmarshal(body, &providerRequest)
+	if err != nil {
+		return fmt.Errorf("unable to unmarshal request body: %v", err)
+	}
+
+	results := make([]externaldata.Item, 0)
+	wg := sync.WaitGroup{}
+	mu := sync.Mutex{}
+
+	for _, image := range providerRequest.Request.Keys {
+		wg.Add(1)
+		go func(image string) {
+			defer wg.Done()
+			routineStartTime := time.Now()
+			logrus.Infof("mutating image %v", image)
+			returnItem := externaldata.Item{
+				Key:   image,
+				Value: image,
+			}
+			defer func() {
+				mu.Lock()
+				results = append(results, returnItem)
+				mu.Unlock()
+			}()
+			parsedReference, err := pkgUtils.ParseSubjectReference(image)
+			if err != nil {
+				errMessage := fmt.Sprintf("failed to mutate image reference %s: %v", image, err)
+				logrus.Error(errMessage)
+				returnItem.Error = errMessage
+				return
+			}
 
-	return sendResponse(&results, "", w, http.StatusOK)
+			if parsedReference.Digest == "" {
+				var selectedStore referrerstore.ReferrerStore
+				for _, store := range server.GetExecutor().ReferrerStores {
+					if store.Name() == server.MutationStoreName {
+						selectedStore = store
+						break
+					}
+				}
+				if selectedStore == nil {
+					errMessage := fmt.Sprintf("failed to mutate image reference %s: could not find matching store: %v", image, err)
```

Comment:
nit: err seems nil here?

Reply:
good catch! it should've been the store name not err